### PR TITLE
Self repair driver

### PR DESCRIPTION
**Pull request by Osmose**
A basic first pass at the self-repair driver.

@mythmon r?

### CODE REVIEW

File: normandy/classifier/api.py [hunk 1]
```diff
@@ -6,7 +6,8 @@
 
 
 class FetchBundle(views.APIView):
-
+    authentication_classes = []
+    permission_classes = []
```
Comment (mythmon) [inline normandy/classifier/api.py]: Why did you add these? Do we need them?

File: normandy/selfrepair/static/js/self_repair_driver.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+ * @param {Object} options.headers - Headers and values to set on the
+ *     request.
+ * @promise {XMLHttpRequest} The request after it has succeeded.
+ * @rejects {XMLHttpRequest} The request after it has failed.
+ */
+function xhr(method, url, options={}) {
+    return new Promise((resolve, reject) => {
+        let req = new XMLHttpRequest();
+        req.addEventListener('load', e => {
+            if (req.status !== 200) {
```
Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_driver.js]: This should be something like `if (req.status >= 400) {`. [201][] and [204][] are totally useful status codes.

[201]: http://http.cat/201
[204]: http://http.cat/204

File: normandy/selfrepair/static/js/self_repair_driver.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ * @promise Resolves once all actions are complete.
+ */
+function runRecipe(recipe) {
+    return Promise.all(recipe.actions.map(loadAction)).then(() => {
+        let recipePromise = Promise.resolve();
+
+        for (let action of recipe.actions) {
+            let func = registeredActions[action.name];
+            recipePromise = recipePromise.then(() => {
+                return func(null, action.arguments);
```
Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_driver.js]: At some point we'll need a better stub instead of `null` here.

File: normandy/selfrepair/static/js/self_repair_driver.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+fetchRecipes().then((recipes) => {
+    let chain = Promise.resolve();
+
+    for (let recipe of recipes) {
+        chain.then(() => {
+            return runRecipe(recipe);
+        });
+    }
+
+    return chain;
```
Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_driver.js]: A silly one liner that does the same thing, if you're into that.

```javascript
recipes.reduce((chain, recipe) => chain.then(() => runRecipe(recipe)), Promise.resolve());
```

File: normandy/classifier/api.py (hunk 1 above)
Comment (Osmose) [inline normandy/classifier/api.py]: We do, because otherwise the view requires a CSRF token due to the default auth/permissions.

File: normandy/selfrepair/static/js/self_repair_driver.js (hunk 3 above)
Comment (Osmose) [inline normandy/selfrepair/static/js/self_repair_driver.js]: Yes, that will be in a later PR

File: normandy/selfrepair/static/js/self_repair_driver.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+function loadAction(action) {
+    return new Promise((resolve, reject) => {
+        if (!registeredActions[action.name]) {
+            let script = document.createElement('script');
+            script.src = action.implementation.url;
+            script.onload = () => {
+                if (!registeredActions[action.name]) {
+                    reject(new Error(`Could not find action with name ${action.name}.`));
+                } else {
+                    resolve();
```
Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_driver.js]: I think this function should resolve with the action that was loaded. That way other things, like `runRecipe` can be simplified a little to not have to deal with the global registry.

File: normandy/selfrepair/static/js/self_repair_driver.js (hunk 4 above)
Comment (Osmose) [inline normandy/selfrepair/static/js/self_repair_driver.js]: I feel like that's super-unreadable. Maybe a helper method for a promise chain would be better, though. Passing that callback to `then` seems like unnecessary noise.

Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_driver.js]: That's fair. I'm not sure what you mean by unnecessary noise you mean though, since `chain.then(() => runRecipe(recipe))` is the same as

```javascript
chain.then(() => {
    return runRecipe(recipe);
});
```

just on fewer lines. Is that the part of the code you meant?

Comment (Osmose) [inline normandy/selfrepair/static/js/self_repair_driver.js]: I meant like maybe we make a helper function that removes it:

```js
promiseChain(list_of_promises);
```

Initially I don't think that'd work because by the time you've generated the promises, they're already executing, but maybe we can figure something out. Maybe it's not worth the time. I'll think about it.

Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_driver.js]: Yeah, it would have to be `promiseChain(list_of_things_that_generate_promises)`. That would probably be `promiseChain(recipes.map(recipe => runRecipe.bind(null, recipe)))`, which isn't very nice.

Maybe something like `promiseSerialMap(runRecipe, recipes)` would be nicer?

Comment (mythmon) [inline normandy/selfrepair/static/js/self_repair_driver.js]: This isn't really an important point. The rest of the patch is good, and we can merge this as is.

File: normandy/classifier/api.py (hunk 1 above)
Comment (mythmon) [inline normandy/classifier/api.py]: Does it make sense to change the defaults? I don't think our API will ever want CSRF.

Comment (Osmose) [inline normandy/classifier/api.py]: I think it's good to make a lack of CSRF the exception rather than the default. It's better to have a view that shouldn't have CSRF accidentally have it then to have the opposite.

**Comment by Osmose** [comment]
Updated!

File: normandy/classifier/api.py (hunk 1 above)
Comment (mythmon) [inline normandy/classifier/api.py]: To be clear, we can make the API specifically CSRF-free, without making that the default for all views.

Comment (Osmose) [inline normandy/classifier/api.py]: I think it's a good thing to do even for the API, but I can change it anyway if you feel strongly about it.
